arwen/host/asyncCall.go

```diff
@@ -195,3 +265,388 @@ func (host *vmHost) computeDataLengthFromArguments(function string, arguments []
+		return err
+	}
+
+	vmInput := runtime.GetVMInput()
+	var asyncCallPosition int
+	var currentContextIdentifier string
+	for contextIdentifier, asyncContext := range asyncInfo.AsyncContextMap {
+		for position, asyncCall := range asyncContext.AsyncCalls {
+			if bytes.Equal(vmInput.CallerAddr, asyncCall.Destination) {
+				asyncCallPosition = position
+				currentContextIdentifier = contextIdentifier
+				break
+			}
+		}
+
+		if len(currentContextIdentifier) > 0 {
+			break
+		}
+	}
+
+	if len(currentContextIdentifier) == 0 {
+		return arwen.ErrCallBackFuncNotExpected
+	}
+
+	// Remove current async call from the pending list
+	currentContextCalls := asyncInfo.AsyncContextMap[currentContextIdentifier].AsyncCalls
+	currentContextCalls[asyncCallPosition] = currentContextCalls[len(currentContextCalls)-1]
+	currentContextCalls[len(currentContextCalls)-1] = nil
+	currentContextCalls = currentContextCalls[:len(currentContextCalls)-1]
+
+	if len(currentContextCalls) == 0 {
+		// call OUR callback for resolving a full context
+		delete(asyncInfo.AsyncContextMap, currentContextIdentifier)
+	}
+
+	// If we are still waiting for callbacks we return
+	if len(asyncInfo.AsyncContextMap) > 0 {
+		return nil
+	}
+
+	_, err = storage.SetStorage(storageKey, nil)
+	if err != nil {
+		return err
+	}
+
+	// Now figure out if we can execute the callback here or different shard
+	if !host.canExecuteSynchronouslyOnDest(asyncInfo.AsyncInitiator.CallerAddr) {
+		err = host.sendStorageCallbackToDestination(asyncInfo.AsyncInitiator)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}
+
+	// The caller is in the same shard, execute it's callback
+	callbackCallInput, err := host.createCallbackContractCallInput(
+		host.Output().GetVMOutput(),
+		asyncInfo.AsyncInitiator.CallerAddr,
+		arwen.CallbackDefault,
+		nil,
+	)
+	if err != nil {
+		return err
+	}
+
+	callbackVMOutput, _, callBackErr := host.ExecuteOnDestContext(callbackCallInput)
+	err = host.processCallbackVMOutput(callbackVMOutput, callBackErr)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+/**
+ * setupAsyncCallsGas sets the gasLimit for each async call with the amount of gas provided by the
+ *  SC developer. The remaining gas is split between the async calls where the developer
+ *  did not specify any gas amount
+ */
+func (host *vmHost) setupAsyncCallsGas(asyncInfo *arwen.AsyncContextInfo) error {
+	gasLeft := host.Metering().GasLeft()
+	gasNeeded := uint64(0)
+	callsWithZeroGas := uint64(0)
+
+	for identifier, asyncContext := range asyncInfo.AsyncContextMap {
+		for index, asyncCall := range asyncContext.AsyncCalls {
+			gasNeeded += asyncCall.ProvidedGas
+			if gasNeeded > gasLeft {
+				return arwen.ErrNotEnoughGas
+			}
+
+			if asyncCall.ProvidedGas == 0 {
+				callsWithZeroGas++
+				continue
+			}
+
+			asyncInfo.AsyncContextMap[identifier].AsyncCalls[index].GasLimit = asyncCall.ProvidedGas
+		}
+	}
+
+	if callsWithZeroGas == 0 {
+		return nil
+	}
+
+	if gasLeft == gasNeeded {
```

Comment:
I've made the condition as you suggested because it seems more obvious what we want there. But `gasLeft` cannot be smaller than `gasNeeded` because of the check from line 574.